x/capability/keeper/keeper.go

```diff
@@ -0,0 +1,244 @@
+	}
+
+	// create new capability with the current global index
+	index := types.IndexFromKey(store.Get(types.KeyIndex))
+	cap := types.NewCapabilityKey(index)
+
+	// update capability owner set
+	if err := sk.addOwner(ctx, cap, name); err != nil {
+		return nil, err
+	}
+
+	// increment global index
+	store.Set(types.KeyIndex, types.IndexToKey(index+1))
+
+	// Set the forward mapping between the module and capability tuple and the
+	// capability name in the in-memory store.
+	memStore.Set(types.FwdCapabilityKey(sk.module, cap), []byte(name))
+
+	// Set the reverse mapping between the module and capability name and the
+	// capability in the in-memory store.
+	memStore.Set(types.RevCapabilityKey(sk.module, name), sk.cdc.MustMarshalBinaryBare(cap))
+
+	logger(ctx).Info("created new capability", "module", sk.module, "name", name)
+	return cap, nil
+}
+
+// AuthenticateCapability attempts to authenticate a given capability and name
+// from a caller. It allows for a caller to check that a capability does in fact
+// correspond to a particular name. The scoped keeper will lookup the capability
+// from the internal in-memory store and check against the provided name. It returns
+// true upon success and false upon failure.
+//
+// Note, the capability's forward mapping is indexed by a string which should
+// contain it's unique memory reference.
+func (sk ScopedKeeper) AuthenticateCapability(ctx sdk.Context, cap types.Capability, name string) bool {
+	memStore := ctx.KVStore(sk.memKey)
+
+	bz := memStore.Get(types.FwdCapabilityKey(sk.module, cap))
+	return string(bz) == name
+}
+
+// ClaimCapability attempts to claim a given Capability and name tuple. This tuple
+// is considered an Owner. It will attempt to add the owner to the persistent
+// set of capability owners for the capability index. If the owner already exists,
+// it will return an error. Otherwise, it will also set a forward and reverse index
+// for the capability and capability name.
+func (sk ScopedKeeper) ClaimCapability(ctx sdk.Context, cap types.Capability, name string) error {
+	// update capability owner set
+	if err := sk.addOwner(ctx, cap, name); err != nil {
+		return err
+	}
+
+	memStore := ctx.KVStore(sk.memKey)
+
+	// Set the forward mapping between the module and capability tuple and the
+	// capability name in the in-memory store.
+	memStore.Set(types.FwdCapabilityKey(sk.module, cap), []byte(name))
+
+	// Set the reverse mapping between the module and capability name and the
+	// capability in the in-memory store.
+	memStore.Set(types.RevCapabilityKey(sk.module, name), sk.cdc.MustMarshalBinaryBare(cap))
+
+	logger(ctx).Info("claimed capability", "module", sk.module, "name", name, "capability", cap.GetIndex())
+	return nil
+}
+
+// GetCapability allows a module to fetch a capability which it previously claimed
+// by name. The module is not allowed to retrieve capabilities which it does not
+// own. If another module claims a capability, the previously owning module will
```

Comment:
THis doesn't seem to be implemented? How do we make sure previously owning module will no longer be able to call it. We aren't deleting mappings from the persistent store or memory store

Especially since capabilities are multi-owner.

cc: @cwgoes

Reply:
This documentation is no longer accurate, perhaps it was copied from an out-of-date line in the spec.